hack/bin/Sbom.hs

```diff
@@ -0,0 +1,376 @@
+- see the corresponding nix code in ./nix
+-}
+
+module Sbom where
+
+import Control.Arrow ((&&&))
+import Data.Aeson
+import Data.Aeson.Key qualified as KM
+import Data.Aeson.KeyMap qualified as KM
+import Data.Aeson.Types (typeMismatch)
+import Data.Bifunctor (first)
+import Data.Bitraversable (bitraverse)
+import Data.ByteString (ByteString)
+import Data.ByteString.Char8 qualified as C8
+import Data.ByteString.Lazy (LazyByteString)
+import Data.ByteString.Lazy qualified as BSL
+import Data.ByteString.Lazy.Char8 qualified as C8L
+import Data.Containers.ListUtils (nubOrd, nubOrdOn)
+import Data.Functor.Identity
+import Data.Map (Map)
+import Data.Map qualified as M
+import Data.Maybe
+import Data.Proxy
+import Data.Text (Text)
+import Data.Text qualified as T
+import Data.Text.IO qualified as T
+import Data.Time.Clock.POSIX
+import Data.Traversable (for)
+import Data.Tree
+import Data.UUID qualified as UUID
+import Data.UUID.V4 qualified as V4
+import Debug.Trace
+import GHC.Generics hiding (Meta)
+import GHC.IsList (IsList (fromList, toList))
+import Numeric.Natural (Natural)
+import Options.Applicative (customExecParser, fullDesc, help, long, prefs, progDesc, showHelpOnEmpty, strOption, value)
+import Options.Applicative qualified as Opt
+import System.Directory
+import System.Process
+
+data License = MkLicense
+  { id :: Maybe Text,
+    name :: Maybe Text
+  }
+  deriving stock (Eq, Ord, Show, Generic)
+  deriving anyclass (FromJSON, ToJSON)
+
+sadSbomMeta :: Text -> Text -> [Text] -> SBomMeta Identity
+sadSbomMeta drvPath outPath directDeps =
+  MkSBomMeta
+    { drvPath = drvPath,
+      outPath = Identity outPath,
+      directDeps = Identity directDeps,
+      description = Nothing,
+      homepage = Nothing,
+      licenseSpdxId = [],
+      name = Nothing,
+      typ = Nothing,
+      urls = [],
+      version = Nothing
+    }
+
+data SBomMeta f = MkSBomMeta
+  { drvPath :: Text,
+    description :: Maybe Text,
+    homepage :: Maybe Text,
+    licenseSpdxId :: [Maybe License],
+    name :: Maybe Text,
+    typ :: Maybe Text,
+    urls :: [Maybe Text],
+    version :: Maybe Text,
+    outPath :: f Text,
+    directDeps :: f [Text]
+  }
+
+deriving stock instance (Eq (f [Text]), Eq (f Text)) => Eq (SBomMeta f)
+
+deriving stock instance (Ord (f [Text]), Ord (f Text)) => Ord (SBomMeta f)
+
+deriving stock instance (Show (f [Text]), Show (f Text)) => Show (SBomMeta f)
+
+type Meta = SBomMeta Proxy
+
+instance FromJSON Meta where
+  parseJSON (Object val) =
+    MkSBomMeta
+      <$> do val .: "drvPath"
+      <*> do val .: "description"
+      <*> do val .: "homepage"
+      <*> do val .: "licenseSpdxId"
+      <*> do val .: "name"
+      <*> do val .: "type"
+      <*> do val .: "urls"
+      <*> do val .: "version"
+      <*> pure Proxy
+      <*> pure Proxy
+  parseJSON invalid = typeMismatch "Object" invalid
+
+type SBom = Map Text (SBomMeta Identity)
+
+type MetaDB = Map Text (SBomMeta Proxy)
```

Comment:
Unfortunately, I don't understand why we need to have `SBomMeta` with `outPath :: Proxy Text` and `outPath :: Proxy Text` :thinking: 

Could you please explain why `outPath :: Text` isn't good enough?

Reply:
Because it’s just not there before. Proxy a is just a weird way of saying Const ()

Reply:
should I replace `Proxy` with `Const ()` to make it more clear?